api/environments/identities/views.py

```diff
@@ -24,29 +28,112 @@
+            filter_kwargs["identifier__exact"] = search_query.replace('"', "")
+        else:
+            filter_kwargs["identifier__icontains"] = search_query
+
+        return filter_kwargs
 
     def get_queryset(self):
         environment = self.get_environment_from_request()
-        queryset = Identity.objects.filter(environment=environment)
 
         search_query = self.request.query_params.get("q")
-        if search_query:
-            if search_query.startswith('"') and search_query.endswith('"'):
-                # Quoted searches should do an exact match just like Google
-                queryset = queryset.filter(
-                    identifier__exact=search_query.replace('"', "")
-                )
-            else:
-                # Otherwise do a fuzzy search
-                queryset = queryset.filter(identifier__icontains=search_query)
+        queryset = Identity.objects.filter(environment=environment)
 
+        if search_query:
+            filter_kwargs = self._get_search_kwargs_django(search_query)
+            queryset = queryset.filter(**filter_kwargs)
         # change the default order by to avoid performance issues with pagination
         # when environments have small number (<page_size) of records
         queryset = queryset.order_by("created_date")
 
         return queryset
 
+    def _get_dynamo_query_kwargs(
+        self,
+        request,
+        environment_api_key: str,
+        previous_last_evaluated_key: str,
+        page_size: int,
+    ) -> dict:
+        search_query = self.request.query_params.get("q")
+
+        filter_expression = Key("environment_api_key").eq(
+            self.kwargs["environment_api_key"]
+        )
+        if search_query:
+            filter_expression = filter_expression & self._get_search_expression_dynamo(
+                search_query
+            )
+
+        dynamo_query_kwargs = {
+            "IndexName": "environment_api_key-identifier-index",
+            "Limit": page_size,
+            "KeyConditionExpression": filter_expression,
+        }
+        if previous_last_evaluated_key:
+            dynamo_query_kwargs.update(
+                ExclusiveStartKey=json.loads(
+                    base64.b64decode(previous_last_evaluated_key)
+                )
+            )
+        return dynamo_query_kwargs
+
+    def list(self, request, *args, **kwargs):
+        environment = self.get_environment_from_request()
+        if not environment.project.enable_dynamo_db:
+            return super().list(request, *args, **kwargs)
+
+        paginator = self.pagination_class()
+        page_size = paginator.get_page_size(request)
+
+        previous_last_evaluated_key = self.request.GET.get("last_evaluated_key")
+        dynamo_query_kwargs = self._get_dynamo_query_kwargs(
+            request, environment.api_key, previous_last_evaluated_key, page_size
+        )
+        response = Identity.query_items_dynamodb(**dynamo_query_kwargs)
+
+        pagination_kwargs = {
```

Comment:
Can we hide some of this in the Pagination class? How will the FE be able to call the next page without changing the structure of the response? At the moment, the pagination is just page number, right? We're essentially introducing a 'limit / offset' pagination format.

Reply:
No, it's using the same structure as postgres db api: i.e: next and previous page url